clientgo/examples: add ToC for examples

ahmetb opened the pull request:
Also add authenticate- prefix to auth samples. This patch could use some
improvement explaining workqueue and TPR examples as I'm not entirely sure.

/assign @caesarxuchao

Signed-off-by: Ahmet Alp Balkan <ahmetb@google.com>

ahmetb committed:
clientgo/examples: add ToC for examples

Also add authenticate- prefix to auth samples.

Signed-off-by: Ahmet Alp Balkan <ahmetb@google.com>

ahmetb commented:
/release-note-none

caesarxuchao commented on hack/verify-staging-imports.sh:
```diff
@@ -79,7 +79,7 @@ if grep -rq '// import "k8s.io/kubernetes/' 'staging/'; then
 fi
 
-for EXAMPLE in vendor/k8s.io/client-go/examples/{in-cluster,out-of-cluster,third-party-resources}; do
+for EXAMPLE in vendor/k8s.io/client-go/examples/{authenticate-in-cluster,authenticate-out-of-cluster,third-party-resources}; do
```
The server url and port is also set up differently, so it's not just authenticate. Maybe call them in/out-cluster-client-configuration?

ahmetb replied:
I think authenticating is fine. The fact that you're pointing server URL/port is an implementation detail of authentication. The name change I made in this case was actually suggested by Jago (on the proposal doc) and I agree that it makes the point clear.

+ Helps with discovery as the letter A is listed first on GitHub.

caesarxuchao replied:
There are ways to configure the client without supplying the authentication information, so authentication doesn't always equal to configuration. So i think "configuration" is slightly better.

> It also helps with discovery as the letter A is listed first on GitHub.

We have your ToC now so we don't depend on the alphabetical order anymore :)

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+# client-go Usage Examples
+
+This directory contains examples that cover examples of various use cases
```
s/examples of//

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+# client-go Usage Examples
+
+This directory contains examples that cover examples of various use cases
+and functionality for Go client library for Kubernetes.
```
s/Go client library for Kubernetes/client-go/

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+This directory contains examples that cover examples of various use cases
+and functionality for Go client library for Kubernetes.
+
+### Authentication
```
Configuration

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+
+- [**Third-party resources**](./third-party-resources): Register a custom
+  resource type with the API, create/update/query this custom type, and write a
+  controller to handle events for this custom resource type.
```
`events` could be misleading, because Kubernetes have an API type called `event`. How about "write a controller drives the cluster state based on the changes to the custom resources."

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+- [**Third-party resources**](./third-party-resources): Register a custom
+  resource type with the API, create/update/query this custom type, and write a
+  controller to handle events for this custom resource type.
+- [**Work queues**](./workqueue): Create a custom controller with a Watch for Pod events,
```
How about "Create a hotloop-free controller with the rate-limited workqueue and the informer framework", i don't think we need to mention "Watch for Pod".

And could you move this example before the tpr one? The latter is more complex.

ahmetb replied:
I don't understand the "Work queue" example just yet but the following 2 concepts are a bit too difficult to write here without any explanation:

- hotloop-free controller
- informer framework.

Is there a better way you think you can explain this example? 

> Create a hotloop-free controller with the rate-limited workqueue and the informer framework

I cannot infer any of this from `workqueue/README.md` which says:

```
This example demonstrates how to write a controller which follows the states
of watched resources.

It demonstrates how to:
 * combine the workqueue with a cache to a full controller
 * synchronize the controller on startup
```

:) Let me know if you have a better explanation and I will change/consolidate.

caesarxuchao replied:
tl;dr
Consider introduce informer in the description of the tpr example:
"The example demonstrates how to write a controller using the [informer](https://godoc.org/k8s.io/client-go/tools/cache#NewInformer)."

And update the description of workqueue example to:
"Combine the workqueue with the informer to write a controller. Workqueue makes sure the controller doesn't hot-loop on failed operations."

---

> hotloop-free controller

I'm referring to this [line](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/examples/workqueue/main.go#L106), this makes sure the controller to exponential backoff before process the item again. Preventing hot-loop is a main feature of the workqueue.

IMO another important functionality of the workqueue is that it allow multiple workers to work in parallel, unfortunately the example doesn't show that.

> informer framework

I mean the "k8s.io/client-go/tools/cache" package, which is also used in the thirdparty resource example, so we don't need to emphasis "informer framework" in the workqueue example.


> combine the workqueue with a cache to a full controller

Here the "cache" means "k8s.io/client-go/tools/cache".

> synchronize the controller on startup

It refers to this [line](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/examples/workqueue/main.go#L126). It's less important. Probably we should copy this line to the tpr example.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,36 @@
+- [**Work queues**](./workqueue): Create a custom controller with a Watch for Pod events,
+  a rate limited Work Queue and an Informer with cache.
+
+## How to run these examples
```
The in-cluster won't work with these steps. And the user needs a kubeconfig ready. I suggest that we direct the readers to follow each example's README to learn how to run them.

ahmetb commented:
@caesarxuchao PTAL again. I have this comment still: https://github.com/kubernetes/kubernetes/pull/46883#discussion_r121024481 I think we need to do a better job summarizing what `workqueue` example does (please suggest one), other than that I addressed your feedback/

caesarxuchao commented:
Also please note the recent changes to the examples. The tpr example is marked as deprecated, the successor is CRD.

ahmetb commented:
@caesarxuchao yeah I incorporated the relevant changes. feel free to take a look again.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/README.md:
```diff
@@ -0,0 +1,26 @@
+### Advanced Concepts
+
+- [**Work queues**](./workqueue): Create a hotloop-free controller with the
+  rate-limited workqueue and the informer framework.
```
Could you make "informer framework" a hyperlink to https://godoc.org/k8s.io/client-go/tools/cache#NewInformer?

ahmetb replied:
Will do. But still my [comment above](https://github.com/kubernetes/kubernetes/pull/46883#discussion_r121024481) stands. This sentence is very confusing to a beginner. We need to simplify.

caesarxuchao commented:
One nit and lgtm.

caesarxuchao commented:
There is a standalone example on informer: https://github.com/kubernetes/kubernetes/pull/44446. I think we can explain informer better after that's merged,

ahmetb commented:
Done.

caesarxuchao commented:
/lgtm
/approve

caesarxuchao commented:
/assign @deads2k for approval

It's mostly doc improvement for client-go, can we add the 1.7 milestone?

deads2k commented:
@caesarxuchao cutting it a little fine, but I think this is the last day of slush


/approve no-issue

ahmetb commented:
@k8s-bot pull-kubernetes-unit test this
looks like a flake.

ahmetb commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

ahmetb commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

ahmetb commented:
It seems like this is stuck at the `pull-kubernetes-federation-e2e-gce` flake.

ahmetb commented:
@caesarxuchao it looks like it's getting blocked on the flakes. how do we get this merged?

caesarxuchao commented:
/retest

caesarxuchao commented:
No rebase needed. We just retry the test. Sigh.

ahmetb commented:
No luck :(

caesarxuchao commented:
Opened an issue: https://github.com/kubernetes/kubernetes/issues/47672

caesarxuchao commented:
/retest
